src/imageio/format/jxl.c

```diff
@@ -0,0 +1,480 @@
+{
+  return "image/jxl";
+}
+
+const char *extension(dt_imageio_module_data_t *data)
+{
+  return "jxl";
+}
+
+int dimension(struct dt_imageio_module_format_t *self, struct dt_imageio_module_data_t *data, uint32_t *width,
+              uint32_t *height)
+{
+  *width = 1073741823;
+  *height = 1073741823;
+  return 1;
+}
+
+int bpp(dt_imageio_module_data_t *data)
+{
+  return 32;
+}
+
+int write_image(struct dt_imageio_module_data_t *data, const char *filename, const void *in_tmp,
+                dt_colorspaces_color_profile_type_t over_type, const char *over_filename, void *exif, int exif_len,
+                int imgid, int num, int total, struct dt_dev_pixelpipe_t *pipe, const gboolean export_masks)
+{
+  // By default error code
+  int out = 1;
+
+  const float *in = (float *)in_tmp;
+  float *pixels = NULL;
+  uint8_t *out_buf = NULL;
+  FILE *out_file = NULL;
+
+#define JXL_ASSERT(code)                                                                                          \
+  {                                                                                                               \
+    JxlEncoderStatus res = code;                                                                                  \
+    if(res != JXL_ENC_SUCCESS)                                                                                    \
+    {                                                                                                             \
+      fprintf(stderr, "JXL call failed with err %d (src/imageio/format/jxl.c#L%d)\n", res, __LINE__);             \
+      goto end;                                                                                                   \
+    }                                                                                                             \
+  }
+
+  dt_imageio_jxl_params_t *params = (dt_imageio_jxl_params_t *)data;
+  int width = params->global.width;
+  int height = params->global.height;
+
+  JxlEncoder *encoder = JxlEncoderCreate(NULL);
+
+  // JxlThreadParallelRunnerDefaultNumWorkerThreads()
+  uint32_t num_threads = JxlResizableParallelRunnerSuggestThreads(width, height);
+  void *runner = JxlThreadParallelRunnerCreate(NULL, num_threads);
+  JXL_ASSERT(JxlEncoderSetParallelRunner(encoder, JxlThreadParallelRunner, runner));
+
+  cmsHPROFILE out_profile = dt_colorspaces_get_output_profile(imgid, over_type, over_filename)->profile;
+
+  cmsUInt32Number size = 0;
+  if(cmsSaveProfileToMem(out_profile, NULL, &size))
+  {
+    unsigned char *buf = malloc(size);
+    if(!buf || !cmsSaveProfileToMem(out_profile, buf, &size)) goto end;
+
+    JXL_ASSERT(JxlEncoderSetICCProfile(encoder, buf, size));
+
+    free(buf);
+  }
+
+  JxlBasicInfo basic_info = {};
+  // fixme: add this in once libjxl 0.6 is released and forced as a dependency
+  // This method is only added in 0.6 but it is required to be called
+  // JXL_ASSERT(JxlEncoderInitBasicInfo(&basic_info));
+  basic_info.have_container = JXL_FALSE;
```

Comment:
To support exif, I am writing my own code to create the bmff container containing the jxl raw codestream and the exif data. 

To get the raw codestream from libjxl we need to tell libjxl not to make its own container.

Sorry that it's taking a while to implement these changes, they require quite a lot of research since all the specs are hidden behind ISO paywalls, a lot of testing, and I have been quite busy recently. Hopefully this PR will be ready soon though...

Reply:
To be honest, I think you efforts would be much more appreciated and have a more meaningful impact of you worked on either a) direct Exif writing API within libjxl or b) BMFF/JXL write support the n exiv2

Reply:
Good advice. Perhaps I will do those in the future. Anyway for now I have finished support within darktable.